### PR TITLE
Feature: mutations

### PR DESCRIPTION
**Pull request by gregorjerse**
## Checklist

* [ ] Update CHANGELOG.rst for each commit separately:
  * Pay attention to write entries under the "Unreleased" section.
  * Mark all breaking changes as "**BACKWARD INCOMPATIBLE:**" and put them
    before non-breaking changes.
  * If a commit modifies a feature listed under "Unreleased" section,
    it might be sufficient to modify the existing CHANGELOG entry from previous
    commit(s).
* [ ] Bump the process version:
  * **MAJOR version (first number)**: Backward incompatible changes (changes
    that break the api/interface). Examples: renaming the input/output, adding
    mandatory input, removing input/output...
  * **MINOR version (middle number)**: add functionality or changes in a
    backwards-compatible manner. Examples: add output field, add non-mandatory
    input parameter, use a different tool that produces same results...
  * **PATCH version (last number)**: changes/bug fixes that do not affect
    the api/interface. Examples: typo fix, change/add warning messages...
* [ ] All inputs are used in process.
* [ ] All output fields have a value assigned to them.

### Processes ###

* Code that was once accepted should be kept like it is. It should be changed
  only when its functionality has changed. Small parts may be changed
  to fix bugs (by changing code only to make it nicer you might introduce a bug).
* Set the optimal number of cores that the tool can use (check if parallelization
  can be performed).
* Description of the tool should be self explanatory with links to the tool's
  homepage and publication.
* Write explanatory input labels and add short command line parameter names
  in square brackets (for example: Maximum length [``--maxlength``]).
* Where possible provide default values for the process input fields.
  If they are provided, then ``required = False`` is not needed.
* Add descriptive error and warning messages.
* Use ``Cmd()`` for each call in the process (call of a tool, a
  Bash command, a Python script or an R script) and handle return codes.
* Add ``self.progress()`` calls where appropriate.

### Commit and PR messages ###

* Each commit should be minimal (i.e. 1 change) and self-contained (including
  tests).
* Mark incomplete PRs with the [WIP] (Work In Progress) tag.
* Do not paste links to private repositories from the public one.

### Tests ###

* Add meaningful names of test files:
  * Genomes: ``genome_species.fa.gz``
  * Annotations: ``annotation_species.gtf.gz``
  * Adapters: ``adapters-name/type-source``.
* Specific files should be saved in a folder with the name of the tested tool.
  This folder should include separate subfolders for inputs and outputs.
* Keep test files' size small. If there is no other option, add the large
  test file to the ``files/large`` directory so it will be tracked with Git LFS.
* Pay attention to edge cases: what will happen if any data is
  missing, if the length of some list is 0, if a string is given where one
  expects an integer, if a file is empty, ...
* When testing workflows, try to verify workflow success by checking one
  of the last steps. Pick the one that is most likely to fail. Also test
  that all steps have 'OK' status.

REID-674 REID-675

### CODE REVIEW

File: resolwe_bio/variants/models.py [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+    variant = models.OneToOneField(
+        Variant, on_delete=models.CASCADE, related_name="annotation"
+    )
+    type = models.CharField(max_length=TYPE_MAX_LENGTH, blank=True, null=True)
+    annotation = models.CharField(max_length=ANNOTATION_MAX_LENGTH)
+    annotation_impact = models.CharField(max_length=ANNOTATION_IMPACT_MAX_LENGTH)
+    gene = models.CharField(max_length=GENE_MAX_LENGTH)
+    protein_impact = models.CharField(max_length=PROTEIN_IMPACT_MAX_LENGTH)
+    feature_id = ArrayField(
+        models.CharField(max_length=FEATURE_ID_MAX_LENGTH), default=list
```
Comment (jkokosar) [inline resolwe_bio/variants/models.py:83]: As a single variant can have an impact on multiple (overlapping) transcripts, we want to retain all available information (list) fields as well as isolate the effect associated with the canonical transcript. Thus, we should have two instances of related fields (one as a list, one as a single value):

The following fields are to be renamed and should be optional and provided as a ``list`` of ``CharField`` fields:
* ``annotations`` 
* ``annotation_impacts`` 
* ``genes``
* ``protein_impacts``
* ``transcript_ids`` 

Additionally, add the following new fields which should be a single element ``CharField`` and also not required. 
* ``canonical_annotation``
* ``canonical_annotation_impact``
* ``canonical_gene``
* ``canonical_protein_impact``
* ``canonical_transcript_id``

Comment (jkokosar) [reply resolwe_bio/variants/models.py:83]: fyi @AnzeLovse

File: resolwe_bio/variants/models.py [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+GENOME_ASSEMBLY_MAX_LENGTH = 20
+CHROMOSOME_MAX_LENGTH = 20
+REFERENCE_MAX_LENGTH = 100
+ALTERNATIVE_MAX_LENGTH = 100
+
+# Metadata
+VARIANT_DATA_SOURCE_MAX_LENGTH = 100
+VARIANT_ANNOTATION_SOURCE_MAX_LENGTH = 100
+ANNOTATION_MAX_LENGTH = 200
+GENOTYPE_MAX_LENGTH = 100
```
Comment (marcellevstek) [inline resolwe_bio/variants/models.py:27]: If this refers to genotypes in format "0/1" or "1/1", it could also be shortened.

File: resolwe_bio/variants/models.py [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+
+    experiment = models.ForeignKey(
+        VariantExperiment,
+        on_delete=models.CASCADE,
+        related_name="variant_calls",
+        null=True,
+        blank=True,
+    )
+
+    # QC data.
```
[the remark this answers is not in the document]
Comment (marcellevstek) [reply resolwe_bio/variants/models.py:140]: @jkokosar @mzganec FYI. In the MSI model, we also use variants that have values other than PASS in the filter column.

[the remark this answers is not in the document]
Comment (jkokosar) [reply resolwe_bio/variants/models.py:140]: @marcellevstek updated comment above.

File: resolwe_bio/variants/models.py (hunk 2 above)
Comment (jkokosar) [reply resolwe_bio/variants/models.py:27]: Good point. Currently, the mutations_table output for GT field is more verbose, for example G/T, T/T, TAA/T. This then gets converted to 0, 1 or 2 format in [VariantsTable](https://github.com/genialis/resolwe-bio-py/blob/master/src/resdk/tables/variant.py#L350). We should do this encoding when writing to DB, and keep the length of this field at max 3 (so that either 0/1 or 1 format is supported).

File: resolwe_bio/tests/unit/test_api.py [hunk 4]
```diff
@@ -2,7 +2,7 @@
 from rest_framework import status
 from rest_framework.test import APIRequestFactory, force_authenticate
 
-from resolwe.flow.models import Collection, Data, DescriptorSchema, Entity, Process
```
Comment (dblenkus) [inline resolwe_bio/tests/unit/test_api.py:5]: Removing the descriptor should go into a separate commit.

File: resolwe_bio/variants/views.py [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+    OrderingFilter,
+)
+from resolwe.rest.serializers import SelectiveFieldMixin
+
+from .models import Variant, VariantAnnotation, VariantCall, VariantExperiment
+
+logger = logging.getLogger(__name__)
+
+
+class VariantSerializer(SelectiveFieldMixin, serializers.ModelSerializer):
```
Comment (dblenkus) [inline resolwe_bio/variants/views.py:28]: Put serializers into a separate file.

File: resolwe_bio/variants/views.py [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+            "genome_assembly",
+            "chromosome",
+            "position",
+            "reference",
+            "alternative",
+            "annotation",
+        ]
+
+
+class VariantFilter(CheckQueryParamsMixin, filters.FilterSet):
```
Comment (dblenkus) [inline resolwe_bio/variants/views.py:47]: And also filters.

File: resolwe_bio/variants/models.py [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+    )
+    clinical_significance = models.CharField(
+        max_length=CLINICAL_SIGNIFICANCE_MAX_LENGTH, blank=True, null=True
+    )
+    # Optional references to the external databases.
+    dbsnp_id = models.CharField(max_length=DBSNP_ID_MAX_LENGTH, blank=True, null=True)
+    clinical_var_id = models.CharField(
+        max_length=CLINICAL_VAR_ID_MAX_LENGTH, blank=True, null=True
+    )
+    data = models.ForeignKey(
```
Comment (dblenkus) [inline resolwe_bio/variants/models.py:96]: Is it ok to delete these data when the Data object is deleted, but the sample may still exist?

File: resolwe_bio/variants/views.py [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+            "clinical_significance": TEXT_LOOKUPS,
+            "dbsnp_id": TEXT_LOOKUPS,
+            "clinical_var_id": TEXT_LOOKUPS,
+        }
+
+
+class VariantAnnotationViewSet(mixins.ListModelMixin, viewsets.GenericViewSet):
+    """VariantAnnotation endpoint."""
+
+    queryset = VariantAnnotation.objects.all()
```
Comment (dblenkus) [inline resolwe_bio/variants/views.py:133]: Do we need some permissions system to access this?

File: resolwe_bio/variants/views.py [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+            "experiment__id": NUMBER_LOOKUPS,
+            "quality": NUMBER_LOOKUPS,
+            "depth": NUMBER_LOOKUPS,
+        }
+
+
+class VariantCallViewSet(mixins.ListModelMixin, viewsets.GenericViewSet):
+    """VariantCall endpoint."""
+
+    queryset = VariantCall.objects.all()
```
Comment (dblenkus) [inline resolwe_bio/variants/views.py:191]: Permissions?

File: resolwe_bio/variants/views.py [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+            "contributor__id": NUMBER_LOOKUPS,
+            "date": DATE_LOOKUPS,
+            "variant_data_source": TEXT_LOOKUPS,
+        }
+
+
+class VariantExperimentViewSet(mixins.ListModelMixin, viewsets.GenericViewSet):
+    """VariantExperiment endpoint."""
+
+    queryset = VariantExperiment.objects.all()
```
Comment (dblenkus) [inline resolwe_bio/variants/views.py:229]: Permissions?

File: tests/urls.py [hunk 11]
```diff
@@ -1,30 +1,38 @@
...
+from resolwe_bio.variants.views import VariantAnnotationViewSet, VariantViewSet
 
 from .routers import SearchRouter
 
-
 EntityViewSet.filterset_class = BioEntityFilter
 
 api_router = routers.DefaultRouter(trailing_slash=False)
-api_router.register(r'sample', EntityViewSet)
+api_router.register(r"sample", EntityViewSet)
```
Comment (dblenkus) [inline tests/urls.py:16]: Can you move reformatting to a separate commit and keep only relevant changes here?

**Review by dblenkus — CHANGES_REQUESTED** [review]

File: pyproject.toml [hunk 12]
```diff
@@ -7,3 +7,10 @@ build-backend = "setuptools.build_meta"
 
 [tool.black]
 target-version = ["py36", "py37", "py38", "py39", "py310", "py311"]
+
+[tool.mypy]
+plugins = ["mypy_django_plugin.main", "mypy_drf_plugin.main"]
```
Comment (dblenkus) [inline pyproject.toml:12]: Add a GHA to check MyPy annotations.

File: resolwe_bio/variants/models.py (hunk 2 above)
Comment (gregorjerse) [reply resolwe_bio/variants/models.py:27]: I changed the length to 3. If is also easy to increase the size in the future if required.

File: resolwe_bio/variants/models.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (gregorjerse) [reply resolwe_bio/variants/models.py:140]: I have synced the model data with the description.

File: resolwe_bio/variants/migrations/0001_initial.py [hunk 13]
```diff
@@ -0,0 +1,192 @@
...
+                    models.AutoField(
+                        auto_created=True,
+                        primary_key=True,
+                        serialize=False,
+                        verbose_name="ID",
+                    ),
+                ),
+                ("quality", models.FloatField()),
+                ("depth", models.PositiveIntegerField()),
+                ("filter", models.CharField(max_length=20)),
```
[the remark this answers is not in the document]
Comment (gregorjerse) [reply resolwe_bio/variants/migrations/0001_initial.py:76]: Migration file was updated to reflect the latest changes.

File: resolwe_bio/variants/models.py [hunk 14]
```diff
@@ -0,0 +1,151 @@
...
+        ]
+
+
+class VariantAnnotation(models.Model):
+    """Describes an annotation of a variant."""
+
+    variant = models.OneToOneField(
+        Variant, on_delete=models.CASCADE, related_name="annotation"
+    )
+    type = models.CharField(max_length=TYPE_MAX_LENGTH, blank=True, null=True)
```
[the remark this answers is not in the document]
Comment (gregorjerse) [reply resolwe_bio/variants/models.py:77]: Fixed. Please check if the descriptions of the choices are OK.

File: resolwe_bio/variants/models.py [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+    )
+    clinical_diagnosis = models.CharField(
+        max_length=CLINICAL_DIAGNOSIS_MAX_LENGTH, blank=True, null=True
+    )
+    clinical_significance = models.CharField(
+        max_length=CLINICAL_SIGNIFICANCE_MAX_LENGTH, blank=True, null=True
+    )
+    # Optional references to the external databases.
+    dbsnp_id = models.CharField(max_length=DBSNP_ID_MAX_LENGTH, blank=True, null=True)
+    clinical_var_id = models.CharField(
```
[the remark this answers is not in the document]
Comment (gregorjerse) [reply resolwe_bio/variants/models.py:93]: Done.

File: tests/urls.py (hunk 11 above)
Comment (gregorjerse) [reply tests/urls.py:16]: Sure.

File: pyproject.toml (hunk 12 above)
Comment (gregorjerse) [reply pyproject.toml:12]: OK.

File: resolwe_bio/tests/unit/test_api.py (hunk 4 above)
Comment (gregorjerse) [reply resolwe_bio/tests/unit/test_api.py:5]: I have removed this file from the commit, since decriptor & descriptor_schema fields are already removed in the master branch.

File: resolwe_bio/variants/models.py [hunk 16]
```diff
@@ -0,0 +1,151 @@
...
+GENE_MAX_LENGTH = 100
+PROTEIN_IMPACT_MAX_LENGTH = 100
+FEATURE_ID_MAX_LENGTH = 200
+FILTER_MAX_LENGTH = 20
+
+
+class Variant(models.Model):
+    """Describe a variant in the database."""
+
+    # Because Django ORM cannot handle composite primary keys, each feature is
```
[the remark this answers is not in the document]
Comment (gregorjerse) [reply resolwe_bio/variants/models.py:43]: Fixed.

File: resolwe_bio/variants/models.py (hunk 1 above)
Comment (gregorjerse) [reply resolwe_bio/variants/models.py:83]: This has been changed: there is a new model for every "row" in the data linked to the `VariantCall` model. Please check if it meets your acceptance criteria.

Comment (jkokosar) [reply resolwe_bio/variants/models.py]: Are individual fields in VariantAnnotationTranscript allowed to be empty? Looks good otherwise.

**Review by dblenkus — APPROVED** [review]
